File: examples/go/cases/tasks/task_create/task_create.api.case.go
package task_create

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"app-protocol/examples/go/core"
	"app-protocol/examples/go/core/shared"
)

type TaskStore interface {
	Read() ([]map[string]any, error)
	Write(value []map[string]any) error
	Reset() error
	Update(updater func(current []map[string]any) ([]map[string]any, error)) ([]map[string]any, error)
}

type API struct {
	Ctx        *core.ApiContext
	domainCase *Domain
}

func NewAPI(ctx *core.ApiContext) *API {
	return &API{
		Ctx:        ctx,
		domainCase: NewDomain(),
	}
}

func (c *API) Handler(input TaskCreateInput) core.ApiResponse[TaskCreateOutput] {
	result := core.ExecuteAPI(
		input,
		c.validate,
		nil,
		nil,
		c.service,
	)

	if result.Success {
		result.StatusCode = 201
		return result
	}

	result.StatusCode = mapErrorCodeToStatus(result.Error)
	return result
}

func (c *API) HandleAny(input any) (any, error) {
	parsed, err := c.parseInput(input)
	if err != nil {
		return c.failure("VALIDATION_FAILED", err, mapErrorCodeToStatus(&shared.AppError{Code: "VALIDATION_FAILED"})), nil
	}

	response := c.Handler(parsed)
	return response, nil
}

func (c *API) Router() core.RouteBinding {
	return core.RouteBinding{
		Method: "POST",
		Path:   "/tasks",
		Handler: func(request core.RouteRequest) (any, error) {
			parsed, err := c.parseInput(request.Body)
			if err != nil {
				return c.failure("VALIDATION_FAILED", err, 400), nil
			}

			return c.Handler(parsed), nil
		},
	}
}

func (c *API) Test() error {
	store, err := c.resolveTaskStore()
	if err != nil {
		return err
	}

	if err := store.Reset(); err != nil {
		return err
	}

	result := c.Handler(TaskCreateInput{
		Title:       "Test task",
		Description: "Created by task_create.api test",
	})

	if !result.Success || result.Data == nil {
		return fmt.Errorf("test: handler should return success")
	}

	if result.StatusCode != 201 {
		return fmt.Errorf("test: successful create must return statusCode 201")
	}

	if result.Data.Task.Status != TaskStatusTodo {
		return fmt.Errorf("test: created task must start in todo")
	}

	persisted, err := store.Read()
	if err != nil {
		return err
	}

	if len(persisted) != 1 {
		return fmt.Errorf("test: created task must be persisted")
	}

	if err := store.Reset(); err != nil {
		return err
	}

	for index := 0; index < 4; index++ {
		response := c.Handler(TaskCreateInput{
			Title: fmt.Sprintf("Concurrent task %d", index+1),
		})
		if !response.Success {
			return fmt.Errorf("test: concurrent creates must all succeed")
		}
	}

	concurrentPersisted, err := store.Read()
	if err != nil {
		return err
	}

	if len(concurrentPersisted) != 4 {
		return fmt.Errorf("test: concurrent creates must persist every task")
	}

	if _, err := c.parseInput(map[string]any{"title": "   "}); err == nil {
		return fmt.Errorf("test: parseInput must reject blank title")
	}

	return nil
}

func (c *API) parseInput(value any) (TaskCreateInput, error) {
	return c.domainCase.ParseInput(value)
}

func (c *API) validate(input TaskCreateInput) error {
	if err := c.domainCase.Validate(input); err != nil {
		return &shared.AppCaseError{
			Code:    "VALIDATION_FAILED",
			Message: err.Error(),
		}
	}

	return nil
}

func (c *API) service(input TaskCreateInput) (TaskCreateOutput, error) {
	taskStore, err := c.resolveTaskStore()
	if err != nil {
		return TaskCreateOutput{}, err
	}

	timestamp := time.Now().UTC().Format(time.RFC3339Nano)
	task := Task{
		ID:          newID(),
		Title:       strings.TrimSpace(input.Title),
		Description: strings.TrimSpace(input.Description),
		Status:      TaskStatusTodo,
		CreatedAt:   timestamp,
		UpdatedAt:   timestamp,
	}

	if _, err := taskStore.Update(func(current []map[string]any) ([]map[string]any, error) {
		record, err := taskToRecord(task)
		if err != nil {
			return nil, err
		}

		return append([]map[string]any{record}, current...), nil
	}); err != nil {
		return TaskCreateOutput{}, err
	}

	if c.Ctx.Logger != nil {
		c.Ctx.Logger.Info("task_create: task persisted", map[string]any{
			"taskId": task.ID,
			"title":  task.Title,
		})
	}

	output := TaskCreateOutput{Task: task}
	if err := c.domainCase.ValidateOutput(output); err != nil {
		return TaskCreateOutput{}, err
	}

	return output, nil
}

func (c *API) resolveTaskStore() (TaskStore, error) {
	if c.Ctx == nil || c.Ctx.Extra == nil {
		return nil, &shared.AppCaseError{
			Code:    "INTERNAL",
			Message: "task_create requires a configured taskStore provider",
		}
	}

	providers, ok := c.Ctx.Extra["providers"].(shared.Dict)
	if !ok {
		return nil, &shared.AppCaseError{
			Code:    "INTERNAL",
			Message: "task_create requires a configured taskStore provider",
		}
	}

	taskStore, ok := providers["taskStore"].(TaskStore)
	if !ok || taskStore == nil {
		return nil, &shared.AppCaseError{
			Code:    "INTERNAL",
			Message: "task_create requires a configured taskStore provider",
		}
	}

	return taskStore, nil
}

func (c *API) failure(code string, err error, statusCode int) core.ApiResponse[TaskCreateOutput] {
	appErr := shared.ToAppError(&shared.AppCaseError{
		Code:    code,
		Message: err.Error(),
	}, code)

	return core.ApiResponse[TaskCreateOutput]{
		AppResult: shared.AppResult[TaskCreateOutput]{
			Success: false,
			Error:   &appErr,
		},
		StatusCode: statusCode,
	}
}

func mapErrorCodeToStatus(err *shared.AppError) int {
	if err == nil {
		return 500
	}

	switch err.Code {
	case "VALIDATION_FAILED":
		return 400
	case "NOT_FOUND":
		return 404
	default:
		return 500
	}
}

func taskToRecord(task Task) (map[string]any, error) {
	encoded, err := json.Marshal(task)
	if err != nil {
		return nil, err
	}

	var record map[string]any
	if err := json.Unmarshal(encoded, &record); err != nil {
		return nil, err
	}

	return record, nil
}

func newID() string {
	buffer := make([]byte, 16)
	if _, err := rand.Read(buffer); err != nil {
		return fmt.Sprintf("%d", time.Now().UnixNano())
	}

	return hex.EncodeToString(buffer)
}

File: examples/kotlin/cases/tasks/task_create/task_create.api.case.kt
package app.protocol.examples.kotlin.cases.tasks.task_create

import app.protocol.examples.kotlin.core.api.ApiContext
import app.protocol.examples.kotlin.core.api.ApiResponse
import app.protocol.examples.kotlin.core.api.BaseApiCase
import app.protocol.examples.kotlin.core.api.RouteBinding
import app.protocol.examples.kotlin.core.shared.AppCaseError
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import kotlin.random.Random
import kotlin.time.Clock
import kotlin.time.ExperimentalTime

private val json = Json { ignoreUnknownKeys = false }

private data class TaskStore(
    val read: suspend () -> List<JsonObject>,
    val write: suspend (List<JsonObject>) -> Unit,
    val reset: (suspend () -> Unit)?,
    val update: suspend (suspend (List<JsonObject>) -> List<JsonObject>) -> List<JsonObject>,
)

private fun mapErrorCodeToStatus(code: String?): Int =
    when (code) {
        "VALIDATION_FAILED" -> 400
        "NOT_FOUND" -> 404
        else -> 500
    }

@OptIn(ExperimentalTime::class)
private fun generateTaskId(): String =
    "task_${Clock.System.now().toEpochMilliseconds()}_${Random.nextLong(0, Long.MAX_VALUE).toString(16)}"

class TaskCreateApi(
    ctx: ApiContext,
) : BaseApiCase<TaskCreateInput, TaskCreateOutput>(ctx) {
    private val domainCase = TaskCreateDomain()

    override val usesService: Boolean = true

    override suspend fun handler(input: TaskCreateInput): ApiResponse<TaskCreateOutput> {
        val result = execute(input)
        return if (result.success) {
            result.copy(statusCode = 201)
        } else {
            result.copy(statusCode = mapErrorCodeToStatus(result.error?.code))
        }
    }

    override fun router(): RouteBinding =
        RouteBinding(
            method = "POST",
            path = "/tasks",
            handler = { request ->
                val body = request.body as? Map<String, Any?>
                handler(
                    TaskCreateInput(
                        title = body?.get("title") as? String ?: "",
                        description = body?.get("description") as? String,
                    ),
                )
            },
        )

    override suspend fun test() {
        val store = resolveTaskStore()
        store.reset?.invoke()

        val result =
            handler(
                TaskCreateInput(
                    title = "Test task",
                    description = "Created by task_create.api test",
                ),
            )
        check(result.success && result.data != null) {
            "test: handler should return success"
        }
        check(result.statusCode == 201) {
            "test: successful create must return statusCode 201"
        }
        check(result.data.task.status == TASK_STATUS_TODO) {
            "test: created task must start in todo"
        }

        val persisted = store.read()
        check(persisted.size == 1) {
            "test: created task must be persisted"
        }

        store.reset?.invoke()

        val concurrentCreates =
            coroutineScope {
                List(4) { index ->
                    async {
                        handler(
                            TaskCreateInput(
                                title = "Concurrent task ${index + 1}",
                            ),
                        )
                    }
                }.awaitAll()
            }
        check(concurrentCreates.all { it.success }) {
            "test: concurrent creates must all succeed"
        }

        val concurrentPersisted = store.read()
        check(concurrentPersisted.size == 4) {
            "test: concurrent creates must persist every task"
        }

        val invalid = runCatching {
            _validate(TaskCreateInput(title = "   "))
        }
        check(invalid.isFailure) {
            "test: _validate must reject blank title"
        }
    }

    override suspend fun _validate(input: TaskCreateInput) {
        try {
            domainCase.validate(input)
        } catch (error: IllegalArgumentException) {
            throw AppCaseError(
                code = "VALIDATION_FAILED",
                message = error.message ?: "task_create validation failed",
            )
        }
    }

    override fun _repository(): Any = resolveTaskStore()

    @OptIn(ExperimentalTime::class)
    override suspend fun _service(input: TaskCreateInput): TaskCreateOutput {
        val taskStore = resolveTaskStore()
        val timestamp = Clock.System.now().toString()
        val task =
            Task(
                id = generateTaskId(),
                title = input.title.trim(),
                description = input.description?.trim()?.takeIf { it.isNotBlank() },
                status = TASK_STATUS_TODO,
                createdAt = timestamp,
                updatedAt = timestamp,
            )

        taskStore.update { tasks: List<JsonObject> ->
            listOf(json.encodeToJsonElement(Task.serializer(), task).jsonObject) + tasks
        }

        ctx.logger.info(
            "task_create: task persisted",
            mapOf(
                "taskId" to task.id,
                "title" to task.title,
            ),
        )

        return TaskCreateOutput(task = task)
    }

    @Suppress("UNCHECKED_CAST")
    private fun resolveTaskStore(): TaskStore {
        val providers = ctx.extra?.get("providers") as? Map<String, Any?>
        val rawStore = providers?.get("taskStore") as? Map<String, Any?>
            ?: throw AppCaseError(
                code = "INTERNAL",
                message = "task_create requires a configured taskStore provider",
            )

        return TaskStore(
            read = rawStore["read"] as? (suspend () -> List<JsonObject>)
                ?: throw AppCaseError("INTERNAL", "task_create taskStore.read is missing"),
            write = rawStore["write"] as? (suspend (List<JsonObject>) -> Unit)
                ?: throw AppCaseError("INTERNAL", "task_create taskStore.write is missing"),
            reset = rawStore["reset"] as? (suspend () -> Unit),
            update = rawStore["update"] as? (suspend (suspend (List<JsonObject>) -> List<JsonObject>) -> List<JsonObject>)
                ?: throw AppCaseError("INTERNAL", "task_create taskStore.update is missing"),
        )
    }
}
